### PR TITLE
Add io.github.Adda0.Stardrop

### PR DESCRIPTION
**Pull request by Adda0**
This PR adds [flatpak support](https://github.com/Adda0/stardrop-flatpak) for an unofficial [fork of Stardrop](https://github.com/Adda0/Stardrop/) Mod Manager for Stardew Valley with custom changes, with the [permission of the upstream Stardrop author](https://github.com/Floogen/Stardrop/issues/79#issuecomment-1862709718), adhering to the GPLv3 licence requirements.

### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I have [built][build] and tested the submission locally.
    - Ignoring the screenshot error as instructed for new flatpaks. But I cannot figure out where to set the developer name so that appstream can find it. Will fix it if I find how to do that.
    ```
    {
    "errors": [
        "appstream-screenshots-not-mirrored-in-ostree"
    ],
    "warnings": [
        "appstream-missing-developer-name"
    ]
    }
    ``` 
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
    - Stardrop allows for storing added mods in any user-specified folder. Therefore, the flatpak has host filesystem access. I am thinking about how to restrict the access further, but that is still a work in progress (hence the commented out different filesystem args). 
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** [permission of the upstream Stardrop author](https://github.com/Floogen/Stardrop/issues/79#issuecomment-1862709718), adhering to the GPLv3 licence requirements.
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*
    - Work in progress to eventually merge the custom changes into the upstream repo and later move this flatpak to the upstream repo, too. 

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: io.github.Adda0.Stardrop.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
+app-id: io.github.Adda0.Stardrop
+runtime: org.freedesktop.Platform
+runtime-version: '22.08'
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.yaml:3]:
```suggestion
runtime-version: '23.08'
```

File: io.github.Adda0.Stardrop.yaml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  - "--socket=x11"
+  - "--share=ipc"
+  - "--share=network"
+  - "--device=dri"
+  # - "--filesystem=~/.local/share/Steam/steamapps/common/Stardew\ Valley/"
+  # - "--filesystem=home"
+  - "--filesystem=host"
+  # - "--filesystem=xdg-download"
+  - "--socket=pulseaudio"
+command: "Stardrop"
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.yaml:76]: This should be on top

File: io.github.Adda0.Stardrop.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.dotnet7
+build-options:
+    append-path: /usr/lib/sdk/dotnet7/bin
+    append-ld-library-path: /usr/lib/sdk/dotnet7/lib
+    env:
+        PKG_CONFIG_PATH: /app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet7/lib/pkgconfig
+modules:
+  # Taken from https://github.com/flathub/com.unity.UnityHub/blob/af171c1ee22994e7ef698004bd44ba7e8bc71cf0/openssl-1.1.yaml.
+  - name: openssl-1.1
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.yaml:14]: openssl 1.1 is no longer developed.

File: io.github.Adda0.Stardrop.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        # x-checker-data:
+        #   type: anitya
+        #   project-id: 20333
+        #   url-template: https://www.openssl.org/source/openssl-$version.tar.gz
+  - name: Stardrop
+    buildsystem: simple
+    build-commands:
+      - "install-sdk.sh"
+      - "dotnet publish -c Release --source ./nuget-sources Stardrop/Stardrop.csproj --runtime linux-x64 --self-contained true"
+      - cp -r /run/build/Stardrop/Stardrop/bin/Release/linux-x64/publish/* /app/bin/
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.yaml:42]: doesn't this make it x86_64 only?

File: io.github.Adda0.Stardrop.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      - install -Dm644 io.github.Adda0.Stardrop.metainfo.xml /app/share/metainfo/io.github.Adda0.Stardrop.metainfo.xml
+      - install -Dm644 io.github.Adda0.Stardrop.desktop /app/share/applications/io.github.Adda0.Stardrop.desktop
+      - install -Dm644 Stardrop_512x512x32.png /app/share/icons/hicolor/512x512/apps/io.github.Adda0.Stardrop.png
+    sources:
+      - type: file
+        path: io.github.Adda0.Stardrop.metainfo.xml
+      - type: file
+        path: io.github.Adda0.Stardrop.desktop
+      - type: file
+        path: Stardrop_512x512x32.png
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.yaml:63]: these do not exist here? Also, desktop file, icon and appdata should come from the upstream, you should submit them.

File: io.github.Adda0.Stardrop.yaml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        #   project-id: 20333
+        #   url-template: https://www.openssl.org/source/openssl-$version.tar.gz
+  - name: Stardrop
+    buildsystem: simple
+    build-commands:
+      - "install-sdk.sh"
+      - "dotnet publish -c Release --source ./nuget-sources Stardrop/Stardrop.csproj --runtime linux-x64 --self-contained true"
+      - cp -r /run/build/Stardrop/Stardrop/bin/Release/linux-x64/publish/* /app/bin/
+    sources:
+      - nuget-sources.json
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.yaml:44]: this does not exist?

File: io.github.Adda0.Stardrop.yaml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+finish-args:
+  # - "--socket=wayland"
+  # - "--socket=fallback-x11"
+  - "--socket=x11"
+  - "--share=ipc"
+  - "--share=network"
+  - "--device=dri"
+  # - "--filesystem=~/.local/share/Steam/steamapps/common/Stardew\ Valley/"
+  # - "--filesystem=home"
+  - "--filesystem=host"
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.yaml:73]: host is too broad

> Stardrop allows for storing added mods in any user-specified folder.

What are these folders?

**Comment by bbhtt** [comment]
> But I cannot figure out where to set the developer name so that appstream can find it. Will fix it if I find how to do that.

In the appdata file add `<developer_name>foo</developer_name>`

File: io.github.Adda0.Stardrop.yaml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+runtime: org.freedesktop.Platform
+runtime-version: '22.08'
+sdk: org.freedesktop.Sdk
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.dotnet7
+build-options:
+    append-path: /usr/lib/sdk/dotnet7/bin
+    append-ld-library-path: /usr/lib/sdk/dotnet7/lib
+    env:
+        PKG_CONFIG_PATH: /app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet7/lib/pkgconfig
```
Comment (hfiguiere) [inline io.github.Adda0.Stardrop.yaml:11]: please use `prepend-pkg-config-path`.

Comment (hfiguiere) [reply io.github.Adda0.Stardrop.yaml:11]: yes the doc for the SDK needs to be updated.

File: io.github.Adda0.Stardrop.yaml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      - "dotnet publish -c Release --source ./nuget-sources Stardrop/Stardrop.csproj --runtime linux-x64 --self-contained true"
+      - cp -r /run/build/Stardrop/Stardrop/bin/Release/linux-x64/publish/* /app/bin/
+    sources:
+      - nuget-sources.json
+
+      - type: archive
+        url: "https://github.com/Adda0/Stardrop/archive/refs/tags/v1.0.0-beta.38-custom_adda.1.tar.gz"
+        sha256: eb1a31fa533ef0d3f4f95d49bb3f7ff1b6692307fb1d0aeb7c21b0b7debcad8d
+      # - type: dir
+      #   path: ../Stardrop-repo/
```
Comment (hfiguiere) [inline io.github.Adda0.Stardrop.yaml:50]:
```suggestion
```

File: io.github.Adda0.Stardrop.yaml [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    sources:
+      - type: file
+        path: io.github.Adda0.Stardrop.metainfo.xml
+      - type: file
+        path: io.github.Adda0.Stardrop.desktop
+      - type: file
+        path: Stardrop_512x512x32.png
+finish-args:
+  # - "--socket=wayland"
+  # - "--socket=fallback-x11"
```
Comment (hfiguiere) [inline io.github.Adda0.Stardrop.yaml:66]:
```suggestion
```

File: io.github.Adda0.Stardrop.yaml (hunk 7 above)
Comment (hfiguiere) [reply io.github.Adda0.Stardrop.yaml:73]: if it's just to install mods, `host` easily too broad.

File: io.github.Adda0.Stardrop.yaml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+      - install -Dm644 io.github.Adda0.Stardrop.desktop /app/share/applications/io.github.Adda0.Stardrop.desktop
+      - install -Dm644 Stardrop_512x512x32.png /app/share/icons/hicolor/512x512/apps/io.github.Adda0.Stardrop.png
+    sources:
+      - type: file
+        path: io.github.Adda0.Stardrop.metainfo.xml
+      - type: file
+        path: io.github.Adda0.Stardrop.desktop
+      - type: file
+        path: Stardrop_512x512x32.png
+finish-args:
```
Comment (hfiguiere) [inline io.github.Adda0.Stardrop.yaml:64]: finish-args is usually put above modules.

File: io.github.Adda0.Stardrop.yaml (hunk 7 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml:73]: That is the problem. Stardrop allows the users to set their own directory where they store their mods. From what I have seen, some put the mods "installed" by Stardrop (extracting user-downloaded archive of the mod) in the game main mods folder (that is the only specific folder I can directly request access to), others put them somewhere in their home directory, others possibly who knows where.

I could possibly say that access to home directory is enough and hope that everyone will have their mods stored under their home directory. However, I think this should be OK even for Steam Decks where the Steam directory is under the home directory somewhere.

Would `home` be strict enough? Or is there a better way to let the user choose an arbitrary folder where Stardrop will then have read-write access with the ability to change the directory afterwards?

File: io.github.Adda0.Stardrop.yaml (hunk 6 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml:44]: It is referring to this list of sources: [nuget-sources.json](https://github.com/Adda0/stardrop-flatpak/blob/main/nuget-sources.json). Should I include this list of sources in this PR already?

File: io.github.Adda0.Stardrop.yaml (hunk 3 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml:14]: Unfortunately, it is. It is a long-standing [issue](https://github.com/Floogen/Stardrop/issues?q=is%3Aissue+libssl) with Stardrop. I would like to update all the dependencies in order to not need libssl1.1 any more, too, but that is still WIP.

File: io.github.Adda0.Stardrop.yaml (hunk 6 above)
Comment (bbhtt) [reply io.github.Adda0.Stardrop.yaml:44]: > Should I include this list of sources in this PR already?

yes

Comment (hfiguiere) [reply io.github.Adda0.Stardrop.yaml:44]: how do you think it finds the file?

File: io.github.Adda0.Stardrop.yaml (hunk 5 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml:63]: Just to clarify, by submitting, you just mean pushing all of them together with `nuget-sources.json` to this PR and the paths here in the manifest file remain the same, i.e., in the current directory. Is that right?

File: io.github.Adda0.Stardrop.yaml (hunk 11 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: Done.

File: io.github.Adda0.Stardrop.yaml (hunk 10 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: Done.

File: io.github.Adda0.Stardrop.yaml (hunk 9 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: Done.

File: io.github.Adda0.Stardrop.yaml (hunk 8 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: Fixed.

File: io.github.Adda0.Stardrop.yaml (hunk 6 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml:51]: File is submitted in this PR.

File: io.github.Adda0.Stardrop.yaml (hunk 5 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: Files are submitted in this PR.

File: io.github.Adda0.Stardrop.yaml (hunk 4 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: I modified the command to hopefully include all architectures where the flatpak might be run.

File: io.github.Adda0.Stardrop.yaml (hunk 2 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: Done.

File: io.github.Adda0.Stardrop.yaml (hunk 1 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: Fixed.

**Comment by Adda0** [comment]
Thank you very much for your reviews and advices. I addressed all the issues. Please, feel free to close those you consider resolved and further comment on those that remain unresolved.

> >  But I cannot figure out where to set the developer name so that appstream can find it. Will fix it if I find how to do that.
>
> In the appdata file add <developer_name>foo</developer_name>

I now have both `<developer><name>foo</name></developer>` and `<developer_name></developer_name>` next to each other. Is that correct?

File: io.github.Adda0.Stardrop.yaml [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        sha256: cf3098950cb4d853ad95c0841f1f9c6d3dc102dccfcacd521d93925208b76ac8
+        x-checker-data:
+          type: anitya
+          project-id: 20333
+          url-template: https://www.openssl.org/source/openssl-$version.tar.gz
+  - name: Stardrop
+    buildsystem: simple
+    build-commands:
+      - "install-sdk.sh"
+      - "dotnet publish -c Release --source ./nuget-sources Stardrop/Stardrop.csproj --runtime linux-x64 --self-contained true"
```
Comment (hfiguiere) [inline io.github.Adda0.Stardrop.yaml:48]: this still looks specific to x86_64.

Comment (Adda0) [reply io.github.Adda0.Stardrop.yaml]: You are right. I missed this one. `dotnet` should pick the correct runtime without explicitly specifying `--runtime` argument, so I just do not specify the runtime now.

File: io.github.Adda0.Stardrop.metainfo.xml [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  <launchable type="desktop-id">io.github.Adda0.Stardrop.desktop</launchable>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://github.com/Adda0/stardrop-flatpak/blob/main/screenshots/main-screenshot-resized.png?raw=true</image>
+    </screenshot>
+    <screenshot>
+      <image>https://github.com/Adda0/stardrop-flatpak/blob/main/screenshots/updates-settings-screenshot-resized.png?raw=true</image>
+    </screenshot>
+    <screenshot>
+      <image>https://github.com/Adda0/stardrop-flatpak/blob/main/screenshots/search-author-mods-screenshot-resized.png?raw=true</image>
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.metainfo.xml:29]: Screenshots should be from a tag/commit

File: io.github.Adda0.Stardrop.yaml [hunk 14]
```diff
@@ -0,0 +1,68 @@
+app-id: io.github.Adda0.Stardrop
+runtime: org.freedesktop.Platform
+runtime-version: '23.08'
+sdk: org.freedesktop.Sdk
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.dotnet7
+build-options:
+    append-path: /usr/lib/sdk/dotnet7/bin
+    append-ld-library-path: /usr/lib/sdk/dotnet7/lib
+    prepend-pkg-config-path: /app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet7/lib/pkgconfig
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.yaml:10]:
```suggestion
    prepend-pkg-config-path: /usr/lib/sdk/dotnet7/lib/pkgconfig
```

File: io.github.Adda0.Stardrop.metainfo.xml (hunk 13 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.metainfo.xml]: Fixed. Linked them from a specific commit.

**Comment by bbhtt** [comment]
bot, build io.github.Adda0.Stardrop

File: io.github.Adda0.Stardrop.metainfo.xml [hunk 15]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+  <id>io.github.adda0.stardrop</id>
```
Comment (bbhtt) [inline io.github.Adda0.Stardrop.metainfo.xml:3]: Is it `adda0` or `Adda0`?

**Comment by bbhtt** [comment]
aarch64 build is failing. Looks like you need to add aarc64 sources in nuget-sources.json

```
MSBuild version 17.7.4+3ebbd7c49 for .NET
  Determining projects to restore...
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Runtime.linux-arm64 with version (= 7.0.14)
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1102:   - Found 1 version(s) in /run/build/Stardrop/nuget-sources [ Nearest version: 8.0.0 ]
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1102: Unable to find package Microsoft.AspNetCore.App.Runtime.linux-arm64 with version (= 7.0.14)
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1102:   - Found 1 version(s) in /run/build/Stardrop/nuget-sources [ Nearest version: 8.0.0 ]
/run/build/Stardrop/Stardrop/Stardrop.csproj : error NU1101: Unable to find package Microsoft.NETCore.App.Host.linux-arm64. No packages exist with this id in source(s): /run/build/Stardrop/nuget-sources
  Failed to restore /run/build/Stardrop/Stardrop/Stardrop.csproj (in 1.91 sec).
FB: host_command_exited_cb 633672 256
```

File: io.github.Adda0.Stardrop.metainfo.xml (hunk 15 above)
Comment (hfiguiere) [reply io.github.Adda0.Stardrop.metainfo.xml:3]: It's case sensitive here and has to match the filename.

```suggestion
  <id>io.github.Adda0.Stardrop</id>
```

**Comment by Adda0** [comment]
The dependencies were added. Can I build flatpak for all architectures locally somehow to test the build for arm64? `flatpak run org.flatpak.Builder` has argument `--arch` which says that the architecture building for must be host compatible, which is not the case on my machine.

File: io.github.Adda0.Stardrop.metainfo.xml (hunk 15 above)
Comment (Adda0) [reply io.github.Adda0.Stardrop.metainfo.xml]: The suggested fix was applied.

**Comment by hfiguiere** [comment]
bot, build io.github.Adda0.Stardrop

**Comment by Adda0** [comment]
I tried to modify Stardrop project to build for arm64, too. However, I don't know how to test the build for flatpak for arm64 locally. Stardrop builds for arm64 correctly, it seems.

**Comment by barthalion** [comment]
bot, build io.github.Adda0.Stardrop

**Pull request closed without merging**